### ISSUE TITLE
ZSH Complaints "bad set of key/value pairs for associative array" for Completion of Service Name for systemctl

### ISSUE THREAD

**Issue by Kipsora** [ohmyzsh/ohmyzsh#8751]
**Describe the bug**

The zsh complaints "_systemctl_unit_state: bad set of key/value pairs for associative array" when I typed in "systemctl enable blu" and then press TAB for completion. 

**To Reproduce**

From a clean install of ohmyzsh:
1. Type "systemctl enable blu" (where I supposed to type "systemctl enable bluetooth.service" as completion)
2. Press TAB
3. See error

**Expected behavior**

I think the complaints "_systemctl_unit_state: bad set of key/value pairs for associative array" should not be occurred here and the command (service name) should be completed.

**Screenshots and/or Recordings**

``` bash
$ systemctl enable blue
_systemctl_unit_state: bad set of key/value pairs for associative array
_systemctl_unit_state: bad set of key/value pairs for associative array
_systemctl_unit_state: bad set of key/value pairs for associative array
```

**Desktop (please complete the following information):**
 - OS / Distro: Arch Linux
 - Latest ohmyzsh Update?: Yes
 - ZSH Version: zsh 5.8 (x86_64-pc-linux-gnu)
 - Terminal emulator: gnome-terminal

**Additional context**

I've also tried to remove the `.zcompdump*` files but it still does not work.

**Comment by mcornella** [comment]
What version of systemd do you run? (`systemctl --version`)

**Comment by Summerer** [comment]
I have faced this issue on `systemd 245 (245.2-2-arch)` and previous builds on 245.

**Comment by mcornella** [comment]
Try to complete the command, then run `typeset -f -t _systemctl_unit_state` and try to complete it again, then post the output shown.

**Comment by Summerer** [comment]
```
+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blue*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blue*'

+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service enabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array

+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blue*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blue*'
+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service enabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array

+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blue*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blue*'
+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service enabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array
```

**Comment by asolopovas** [comment]
+1

**Comment by Kipsora** [comment]
> The `_systemctl` completion file is distributed by systemd, what version of systemd do you run? (`systemctl --version`)

Thanks for the quick reply! I have the similar output to Summerer:
```bash
$ typeset -f -t _systemctl_unit_state
$ systemctl --version 
systemd 245 (245.2-2-arch)
+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
$ systemctl enable blu+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blu*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blu*'

+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service disabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array

+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blu*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blu*'
+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service disabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array

+_systemctl_unit_state:0> typeset -gA _sys_unit_state
+_systemctl_unit_state:0> _sys_unit_state=+_systemctl_unit_state:0> __systemctl list-unit-files 'blu*'
+__systemctl:2> systemctl --full --no-legend --no-pager list-unit-files 'blu*'
+_systemctl_unit_state:0> _sys_unit_state=( bluetooth-mesh.service disabled disabled bluetooth.service disabled disabled bluetooth.target static disabled ) _systemctl_unit_state: bad set of key/value pairs for associative array
```

**Comment by mcornella** [comment]
Looks like systemd 245 probably changed the output of `systemctl list-unit-files` to add a second column, which is what's being added to the `_sys_unit_state` associative array. This is the output I get in 242 (note there are only 2 columns instead of the 3 you get):
```
$ systemctl --full --no-legend --no-pager list-unit-files 'blu*'
bluetooth.service enabled
bluetooth.target  static
```
This is a bug in either [the completion script](https://github.com/systemd/systemd/blob/ea500ac513cf51bcb79a5666f1519499d029428f/shell-completion/zsh/_systemctl.in#L223-L224) or the `list-unit-files` subcommand, which changed the output without maintaining backward compatibility. I suggest you open an issue on the systemd repo.

**Comment by Summerer** [comment]
systemd master got a workaround/fix applied. It involves the flag `--no-legend` but I didn't look into the details.
see [systemd issue 15077](https://github.com/systemd/systemd/issues/15077)

**Comment by mcornella** [comment]
While a fix makes it into a new release, you could redefine the `_systemctl_unit_state` function to only use the first two columns:
```zsh
_systemctl_unit_state() {
  typeset -gA _sys_unit_state
  _sys_unit_state=( $(__systemctl list-unit-files "$PREFIX*" | awk '{print $1, $2}') ) }
```

**Comment by lucki1000** [comment]
> While a fix makes it into a new release, you could redefine the `_systemctl_unit_state` function to only use the first two columns:
> 
> ```shell
> _systemctl_unit_state() {
>   typeset -gA _sys_unit_state
>   _sys_unit_state=( $(__systemctl list-unit-files "$PREFIX*" | awk '{print $1, $2}') ) }
> ```

Hi,

I have the same issue on Arch Arm where can i redefine this function?

**Comment by Borderliner** [comment]
> > While a fix makes it into a new release, you could redefine the `_systemctl_unit_state` function to only use the first two columns:
> > ```shell
> > _systemctl_unit_state() {
> >   typeset -gA _sys_unit_state
> >   _sys_unit_state=( $(__systemctl list-unit-files "$PREFIX*" | awk '{print $1, $2}') ) }
> > ```
> 
> Hi,
> 
> I have the same issue on Arch Arm where can i redefine this function?

I think copy/pasting the above function into the bottom of `~/.zshrc`  does the job.

**Comment by rishabhdeepsingh** [comment]
Hey guys, I am also facing the same issue, on my MAC terminal. how do I fix it? 
I tried adding the above function to my `.zshrc` but didn't work.

**Comment by mcornella** [comment]
If this doesn't happen with systemctl completion you should open a new issue.

**Comment by huyhoang8398** [comment]
I still having this issue on Ubuntu 20.04 with zsh 5.8 (x86_64-ubuntu-linux-gnu)

**Comment by mcornella** [comment]
The fix hasn't been released yet (will make it into zsh 5.9), look at the above workaround.
